crates/core/src/factories/stub_transport/test.rs

```diff
@@ -0,0 +1,348 @@
+type V = Box<dyn Fn(Url, bytes::Bytes) -> K2Result<()> + 'static + Send + Sync>;
+
+struct TrackHnd {
+    track: Mutex<Vec<Track>>,
+    preflight_gather_outgoing: G,
+    preflight_validate_incoming: V,
+}
+
+impl std::fmt::Debug for TrackHnd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        (*self.track.lock().unwrap()).fmt(f)
+    }
+}
+
+impl TxBaseHandler for TrackHnd {
+    fn new_listening_address(&self, this_url: Url) {
+        self.track.lock().unwrap().push(Track::ThisUrl(this_url));
+    }
+
+    fn peer_connect(&self, peer: Url) -> K2Result<()> {
+        self.track.lock().unwrap().push(Track::Connect(peer));
+        Ok(())
+    }
+
+    fn peer_disconnect(&self, peer: Url, reason: Option<String>) {
+        self.track
+            .lock()
+            .unwrap()
+            .push(Track::Disconnect(peer, reason));
+    }
+}
+
+impl TxHandler for TrackHnd {
+    fn preflight_gather_outgoing(
+        &self,
+        peer_url: Url,
+    ) -> K2Result<bytes::Bytes> {
+        (self.preflight_gather_outgoing)(peer_url)
+    }
+
+    fn preflight_validate_incoming(
+        &self,
+        peer_url: Url,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        (self.preflight_validate_incoming)(peer_url, data)
+    }
+}
+
+impl TxSpaceHandler for TrackHnd {
+    fn recv_space_notify(&self, peer: Url, space: SpaceId, data: bytes::Bytes) {
+        self.track
+            .lock()
+            .unwrap()
+            .push(Track::SpaceRecv(peer, space, data));
+    }
+}
+
+impl TxModuleHandler for TrackHnd {
+    fn recv_module(
+        &self,
+        peer: Url,
+        space: SpaceId,
+        module: String,
+        data: bytes::Bytes,
+    ) {
+        self.track
+            .lock()
+            .unwrap()
+            .push(Track::ModRecv(peer, space, module, data));
+    }
+}
+
+impl TrackHnd {
+    pub fn new() -> Arc<Self> {
+        Self::new_preflight(
+            Box::new(|_| Ok(bytes::Bytes::new())),
+            Box::new(|_, _| Ok(())),
+        )
+    }
+
+    pub fn new_preflight(g: G, v: V) -> Arc<Self> {
+        Arc::new(Self {
+            track: Mutex::new(Vec::new()),
+            preflight_gather_outgoing: g,
+            preflight_validate_incoming: v,
+        })
+    }
+
+    pub fn url(&self) -> Url {
+        for t in self.track.lock().unwrap().iter() {
+            if let Track::ThisUrl(url) = t {
+                return url.clone();
+            }
+        }
+        panic!("no url found");
+    }
+
+    pub fn assert_connect(&self, peer: &Url) {
+        for t in self.track.lock().unwrap().iter() {
+            if let Track::Connect(u) = t {
+                if u == peer {
+                    return;
+                }
+            }
+        }
+        panic!(
+            "matching connect entry not found {peer}, {:#?}",
+            self.track.lock().unwrap()
+        );
+    }
+
+    pub fn disconnect(&self, peer: &Url, reason: Option<&str>) -> bool {
```

Comment:
```suggestion
    pub fn assert_disconnect(&self, peer: &Url, reason: Option<&str>) -> bool {
```

Reply:
This *used* to have an assert in it, but one of my tests needed it to not panic, so it now returns a bool, and doesn't assert internally.

Reply:
Maybe I should actually remove the asserts from the other helpers?

Reply:
I see, yes that's probably a good idea. Maybe something like "check_xxx" instead of "assert_xxx" and have them return what they find for assertion?